internal/provider/string.go

```diff
@@ -257,3 +323,78 @@ func generateRandomBytes(charSet *string, length int) ([]byte, error) {
 func readNil(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return nil
 }
+
+func resourceStateUpgradeAddNumeric(resourceName string) func(_ context.Context, rawState map[string]interface{}, _ interface{}) (map[string]interface{}, error) {
+	return func(_ context.Context, rawState map[string]interface{}, _ interface{}) (map[string]interface{}, error) {
+		if rawState == nil {
+			return nil, fmt.Errorf("resource %s state upgrade failed, state is nil", resourceName)
```

Comment:
Does Terraform CLI not already make it clear which resource address failed, which would inherently include the resource name? If not, maybe we should raise a feature request upstream so adding the resource name, which may not be the most helpful in large configurations, is not necessary in provider code.

Reply:
I've just checked this. Terraform CLI _does_ make it clear which resource address failed. For instance,
```bash
│ Error: state upgrade failed, state is nil
│ 
│   with random_password.password,
│   on resource.tf line 1, in resource "random_password" "password":
│    1: resource "random_password" "password" {
```

Consequently, I have removed usage of the resource name from the error messages in and refactored to replace `resourceStateUpgradeAddNumeric` func with `resourcePasswordStringStateUpgradeV1`.